video/out/wayland_common.c

```diff
@@ -1297,3 +1313,31 @@ bool vo_wayland_config (struct vo *vo, uint32_t flags)
 
     return true;
 }
+
+void vo_wayland_request_frame(struct vo *vo, void *data, vo_wayland_frame_cb cb)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+    wl->frame.data = data;
+    wl->frame.function = cb;
+    MP_DBG(wl, "restart frame callback\n");
+    frame_callback(wl, NULL, 0);
+}
+
+bool vo_wayland_wait_frame(struct vo *vo)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+
+    if (!wl->frame.callback)
+        return false;
+
+    // Reduce timeout if there are more than 3 pending waits.
+    // Most likely the mpv isn't receiving frame callbacks then.
+    // So don't timeout too long to avoid AV desync.
+    // 10ms matches the time specified in eglSwapInterval(0) mesa patch for wayland.
+    const int timeout = (wl->frame.pending > 3 ? 10 : 100); // ms
+    const int64_t start = mp_time_us();
+    while (wl->frame.pending > 0 && vo_wayland_poll(vo, timeout) > 0 && (mp_time_us() - start) / 1000 < timeout);
+    const bool was_pending = wl->frame.pending;
+    wl->frame.pending = (wl->frame.pending >= 4 ? 4 : wl->frame.pending + 1);
+    return !was_pending;
+}
```

Comment:
This might be better tied to the ms needed to draw video fullspeed. Does mpv have this information anywhere?